File: world_simulator_cpp/Antelope.cpp
#include<iostream>
#include<time.h>
#include "enumList.h"
#include "World.h"
#include "Antelope.h"
using namespace std;

Antelope::Antelope(World* currentWorld, int X_coordinate, int Y_coordinate) : Animal(currentWorld, X_coordinate, Y_coordinate, 4, 4, 'A') {

}

void Antelope::Collision(Organism* predator, Organism* prey) {
	if (predator->GetToken() != prey->GetToken()) {
		cout << predator->GetToken() << " attacks " << prey->GetToken() << " on x: " << prey->GetX() << " y: " << prey->GetY() << endl;
		int escape = rand() % 2;
		if (escape == 0) {
			if (prey->GetStrength() > predator->GetStrength()) {
				this->GetWorld()->DelFromWorldMap(predator);
				this->GetWorld()->DelFromOrganismList(predator);
				delete predator;
			}
			else {
				predator->ChangeLocation(prey->GetX(), prey->GetY());
				this->GetWorld()->DelFromWorldMap(prey);
				this->GetWorld()->DelFromOrganismList(prey);
				predator->Draw();
			}
		}
		else {
			predator->ChangeLocation(prey->GetX(), prey->GetY());
			predator->Draw();
			bool freeSpace = false;
			int new_x = prey->GetX(), new_y = prey->GetY();
			if (!this->currentWorld->IsOutsideBorder(prey->GetX() - 1, prey->GetY())) {
				if (!this->currentWorld->IsFieldOccupied(prey->GetX() - 1, prey->GetY())) {
					new_x--;
					freeSpace = true;
				}
			}
			if (!this->currentWorld->IsOutsideBorder(prey->GetX() + 1, prey->GetY()) && !freeSpace) {
				if (!this->currentWorld->IsFieldOccupied(prey->GetX() + 1, prey->GetY())) {
					new_x++;
					freeSpace = true;
				}
			}
			if (!this->currentWorld->IsOutsideBorder(prey->GetX(), prey->GetY() - 1) && !freeSpace) {
				if (!this->currentWorld->IsFieldOccupied(prey->GetX(), prey->GetY() - 1)) {
					new_y--;
					freeSpace = true;
				}
			}
			if (!this->currentWorld->IsOutsideBorder(prey->GetX(), prey->GetY() + 1) && !freeSpace) {
				if (!this->currentWorld->IsFieldOccupied(prey->GetX(), prey->GetY() + 1)) {
					new_y++;
					freeSpace = true;
				}
			}
			if (freeSpace) {
				prey->ChangeLocation(new_x, new_y);
				prey->Draw();
				predator->Draw();
				cout << "A escaped to x: " << prey->GetX() << " y: " << prey->GetY() << endl;
			}
		}
	}
	else {
		if (predator->GetAge() > 3 && prey->GetAge() > 3)
			this->currentWorld->GiveBirth(predator);
	}
}
void Antelope::Move() {
	bool LegalMove = false;
	int x, y;
	while (!LegalMove) {
		x = 0;
		y = 0;
		int direction = rand() % 4;
		if (direction == DOWN) {
			y += 2;
			if (!currentWorld->IsOutsideBorder(x + GetX(), y + GetY()))
				LegalMove = true;
		}
		else if (direction == RIGHT) {
			x += 2;
			if (!currentWorld->IsOutsideBorder(x + GetX(), y + GetY()))
				LegalMove = true;
		}
		else if (direction == UP) {
			y -= 2;
			if (!currentWorld->IsOutsideBorder(x + GetX(), y + GetY()))
				LegalMove = true;
		}
		else {
			x -= 2;
			if (!currentWorld->IsOutsideBorder(x + GetX(), y + GetY()))
				LegalMove = true;
		}
		if (LegalMove) {
			if (this->currentWorld->IsFieldOccupied(x + this->GetX(), y + this->GetY())) {
				this->currentWorld->GetField(x + this->GetX(), y + this->GetY())->Collision(this, this->currentWorld->GetField(x + this->GetX(), y + this->GetY()));
			}
			else {
				ChangeLocation(x + GetX(), y + GetY());
				cout << this->GetToken() << " moves to x: " << this->GetX() << " y: " << this->GetY() << endl;
			}

		}
	}
}

File: simulator/Organisms/Animals/Antelope.py
from simulator.Organisms.Animal import Animal
import random


class Antelope(Animal):
    def __init__(self, current_world, x_coordinate, y_coordinate):
        super().__init__(current_world, x_coordinate, y_coordinate, 4, 4, True, 'A', 'simulator\\images\\antelope.png')
        self.set_move_dist(2)

    def collision(self, predator):
        text = ""
        if predator.get_token() != self.get_token():
            text += str(predator.get_token()) + " attacks " + str(self.get_token()) + " "
            if random.randint(0, 1) == 0:
                if (self.get_strength() > predator.get_strength()
                        or (self.get_strength() == predator.get_strength() and self.get_age() > predator.get_age())):
                    text += str(self.get_token()) + " wins"
                    self.get_world().add_log(text)
                    self.get_world().del_organism(predator)
                else:
                    text += str(predator.get_token()) + " wins"
                    self.get_world().add_log(text)
                    tmp_x = self.get_x()
                    tmp_y = self.get_y()
                    self.get_world().del_organism(self)
                    predator.new_location(tmp_x, tmp_y)
            else:
                tmp_x = self.get_x()
                tmp_y = self.get_y()
                free_space = False
                new_x = self.get_x()
                new_y = self.get_y()
                if not self.get_world().is_outside(self.get_x() - 1, self.get_y()):
                    if not self.get_world().is_occupied(self.get_x() - 1, self.get_y()):
                        new_x -= 1
                        free_space = True
                if not self.get_world().is_outside(self.get_x() + 1, self.get_y()) and not free_space:
                    if not self.get_world().is_occupied(self.get_x() + 1, self.get_y()):
                        new_x += 1
                        free_space = True
                if not self.get_world().is_outside(self.get_x(), self.get_y() - 1) and not free_space:
                    if not self.get_world().is_occupied(self.get_x(), self.get_y() - 1):
                        new_y -= 1
                        free_space = True
                if not self.get_world().is_outside(self.get_x(), self.get_y() + 1) and not free_space:
                    if not self.get_world().is_occupied(self.get_x(), self.get_y() + 1):
                        new_y += 1
                        free_space = True
                if free_space:
                    text += "A escapes"
                    self.get_world().add_log(text)
                    self.new_location(new_x, new_y)
                    predator.new_location(tmp_x, tmp_y)
                    predator.get_world().set_field(tmp_x, tmp_y, predator)
        elif self.get_age() > 3 and predator.get_age() > 3:
            self.get_world().new_organism(self)
